models/repo_generate.go

```diff
@@ -36,8 +41,148 @@ func (gro GenerateRepoOptions) IsValid() bool {
+type GiteaTemplate struct {
+	Path    string
+	Content []byte
+
+	globs []glob.Glob
+}
+
+// Globs parses the .gitea/template globs or returns them if they were already parsed
+func (gt GiteaTemplate) Globs() []glob.Glob {
+	if gt.globs != nil {
+		return gt.globs
+	}
+
+	gt.globs = make([]glob.Glob, 0)
+	lines := strings.Split(string(util.NormalizeEOL(gt.Content)), "\n")
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+		g, err := glob.Compile(line, '/')
+		if err != nil {
+			log.Info("Invalid glob expression '%s' (skipped): %v", line, err)
+			continue
+		}
+		gt.globs = append(gt.globs, g)
+	}
+	return gt.globs
+}
+
+func checkGiteaTemplate(tmpDir string) (*GiteaTemplate, error) {
+	gtPath := filepath.Join(tmpDir, ".gitea", "template")
+	if _, err := os.Stat(gtPath); os.IsNotExist(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	content, err := ioutil.ReadFile(gtPath)
+	if err != nil {
+		return nil, err
+	}
+
+	gt := &GiteaTemplate{
+		Path:    gtPath,
+		Content: content,
+	}
+
+	return gt, nil
+}
+
+func generateRepoCommit(e Engine, repo, templateRepo, generateRepo *Repository, tmpDir string) error {
+	commitTimeStr := time.Now().Format(time.RFC3339)
+	authorSig := repo.Owner.NewGitSig()
+
+	// Because this may call hooks we should pass in the environment
+	env := append(os.Environ(),
+		"GIT_AUTHOR_NAME="+authorSig.Name,
+		"GIT_AUTHOR_EMAIL="+authorSig.Email,
+		"GIT_AUTHOR_DATE="+commitTimeStr,
+		"GIT_COMMITTER_NAME="+authorSig.Name,
+		"GIT_COMMITTER_EMAIL="+authorSig.Email,
+		"GIT_COMMITTER_DATE="+commitTimeStr,
+	)
+
+	// Clone to temporary path and do the init commit.
+	templateRepoPath := templateRepo.repoPath(e)
+	if err := git.Clone(templateRepoPath, tmpDir, git.CloneRepoOptions{
+		Depth: 1,
+	}); err != nil {
+		return fmt.Errorf("git clone: %v", err)
+	}
+
+	if err := os.RemoveAll(path.Join(tmpDir, ".git")); err != nil {
+		return fmt.Errorf("remove git dir: %v", err)
+	}
+
+	// Variable expansion
+	gt, err := checkGiteaTemplate(tmpDir)
+	if err != nil {
+		return fmt.Errorf("checkGiteaTemplate: %v", err)
+	}
+
+	if err := os.Remove(gt.Path); err != nil {
+		return fmt.Errorf("remove .giteatemplate: %v", err)
+	}
+
+	// Avoid walking tree if there are no globs
+	if len(gt.Globs()) > 0 {
+		tmpDirSlash := strings.TrimSuffix(filepath.ToSlash(tmpDir), "/") + "/"
+		if err := filepath.Walk(tmpDirSlash, func(path string, info os.FileInfo, walkErr error) error {
+			if walkErr != nil {
+				return walkErr
+			}
+
+			if info.IsDir() {
+				return nil
+			}
+
+			base := strings.TrimPrefix(filepath.ToSlash(path), tmpDirSlash)
+			for _, g := range gt.Globs() {
+				if g.Match(base) {
+					content, err := ioutil.ReadFile(path)
+					if err != nil {
+						return err
+					}
+
+					if err := ioutil.WriteFile(path,
+						[]byte(generateExpansion(string(content), templateRepo, generateRepo)),
+						0644); err != nil {
```

Comment:
[Git supports](https://dzone.com/articles/tracking-file-permissions-in-git) the executable flag (e.g. `0755`) for the file owner. May be we should match the template permissions on the updated file? If we won't (I don't think it's crucial), could you leave a TODO comment here?

Note: maybe since you're not creating the file from scratch it will retain its current permissions.

Reply:
Now that we're truncating a file instead of making a new one, this shouldn't change any perms.